module3/lessons/api_authorization.md

```diff
@@ -0,0 +1,502 @@
+
+#### Control for Admin User
+
+Now that we have implemented roles for our users, we can control who has access to perform certain actions:
+
+```ruby
+class Api::V1::SongsController < ApplicationController
+  before_action :authenticate_user
+
+  ...
+
+  def create
+    if @current_user.admin?
+      render json: Song.create(song_params) 
+    else
+      render status: :method_not_allowed
+    end
+  end
+
+  ...
+
+  private
+  ...
+
+end
+```
+
+At this point, we have satisfied the user stories outlined at the beginning of the lesson!
+
+<section class='call-to-action'>
+
+Solidify your learning by implementing the following:
+
+<section class='dropdown'>
+
+### 🫑 Control Songs#delete
+
+Update your tests and implement code to satisfy this user story:
+
+```
+As an admin user
+When I make an api request (with appropriate headers) to
+  DELETE "/api/v1/songs/1"
+Then I receive a successful response
+  And the song is removed from the database
+```
+
+</section>
+
+<section class='dropdown'>
+
+### 🌶️ Control Songs#update
+
+Update your tests and implement code to satisfy this user story:
+
+```
+When I make an api request (with appropriate headers) to
+  PATCH "/api/v1/songs/1" {'title':'New Song Title'}
+Then I see a json response with the song's updated information
+  And the song is updated in the database
+```
+</section>
+
+<section class='dropdown'>
+
+### 🌶️🌶️ DRY your SongsController
+
+At this point, you have a lot of repeated code in your SongsController. Clean up this code so that you are using a callback to control admin access to the CUD related controller actions.
+
+</section>
+
+<section class='dropdown'>
+
+### 🌶️🌶️ SRP Refactor
+
+Think about the responsibility of your SongsController.  Are there any responsibilities that could be abstracted away to new or existing classes?  If so, try to make those refactors!
+
+</section>
+
+</section>
+
+## What About Sessions?
+In the pre-work for this lesson, you learned about sessions - they help an application maintain state for a user by passing session information back and forth in the request and response headers.  This allows us to **stay logged in** on websites like Amazon, Gmail, etc...
+
+But, for API consumption, the session is largely irrelevant because we expect the consumers of our APIs (whether human or application) to tell us who they are with **each request**.  
+
+## Checks for Understanding (aka possible interview questions!)
+* Describe the difference between authentication and authorization.
+* Outline the design of an ebay copycat application:
+    * what resources might exist?
+    * what routes might exist?
+    * How many roles would you need?
+    * what endpoints would be useful?
+    * how might name-spacing be used to structure the design?
+* Define what a 'callback' is in rails; and provide examples of when one would be used.
```

Comment:
This is a great lesson. It's beefy but I think the role exploration is a nice addition and great for them to know for consultancy and beyond.